src/concurrent-queue.js

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright 2020 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import Async from 'async';
+import { v4 as uuid } from 'uuid';
+
+import UploadOptionsBase from './upload-options-base';
+
+/**
+ * Creates a new "batch" that has a given number of items in it. When all
+ * items in the batch have been processed, the given callback will be invoked.
+ * @param {number} totalCount Total number of items in the batch.
+ * @param {function} callback Invoked without arguments when all items in the
+ *  batch have processed.
+ * @returns {string} The ID of the new batch.
+ */
+function createBatch(totalCount, callback) {
+    const batchId = uuid();
+    this.batchCounts[batchId] = {
+        totalCount,
+        totalProcessed: 0,
+        callback,
+    };
+    return batchId;
+}
+
+/**
+ * Increments the total number of items that have been processed in a batch.
+ * If all items in the batch are processed, the method will also invoke the
+ * batch's callback.
+ * @param {string} batchId Batch whose processed count will be incremented.
+ */
+function updateBatch(batchId) {
+    this.batchCounts[batchId].totalProcessed += 1;
```

Comment:
There are actually some [oddities](https://eslint.org/docs/rules/no-plusplus) with unary operators in javascript that make me (though admittedly inconsistently :) ) avoid using them. That's why I've purposely used `+= 1` here instead of the `++`.

Reply:
(Note: Sure. But here you have the semicolon as `end-of-line`, and we're not in a loop here. So we should be fine.)